src/main/kotlin/com/exactpro/th2/codec/AbstractSyncCodec.kt

```diff
@@ -77,23 +77,14 @@ abstract class AbstractSyncCodec(
             }
 
             CompletableFuture.allOf(*messageGroupFutures).whenComplete { _, _ ->
-                messageGroupFutures
-                    .filter { it.get() != null && it.get()!!.messagesCount > 0 }
-                    .forEach {
-                        it.get()!!.apply {
-                            if (checkResult(this)) {
-                                resultBuilder.addGroups(this)
-                            }
-                        }
-                    }
+                messageGroupFutures.forEach { it.get()?.run(resultBuilder::addGroups) }
             }.get()
         } else {
-            groupBatch.groupsList
-                .asSequence()
-                .filter { it.messagesCount > 0 }
-                .mapIndexedNotNull(::runProcessMessageGroup)
-                .filter(::checkResult)
-                .forEach(resultBuilder::addGroups)
+            groupBatch.groupsList.forEachIndexed { index, group ->
+                if (group.messagesCount == 0) return@forEachIndexed
+                val processedGroup = runProcessMessageGroup(index, group) ?: return@forEachIndexed
+                if (checkResult(processedGroup)) resultBuilder.addGroups(processedGroup)
```

Comment:
> if (group.messagesCount == 0)
if (checkResult(processedGroup))

These checks are redundant since they will be done by `runProcessMessageGroup`

```suggestion
                runProcessMessageGroup(index, group)?.run(resultBuilder::addGroups)
```